File: 0710-Random Pick with Blacklist/0710-Random Pick with Blacklist.cpp
class Solution {
public:
    Solution(int N, vector<int> blacklist) {
        wlen = N - blacklist.size();
        for (int b : blacklist) {
            table[b] = -1;
        }
        
        int mapped = N - 1;
        for (int b : blacklist) {
            if (b < wlen) {
                while (table.count(mapped)) {
                    --mapped;
                }
                table[b] = mapped--;
            }
        }
    }
    
    int pick() {
        int index = rand() % wlen;
        return table.count(index) ? table[index] : index;
    }
    
private:
    unordered_map<int, int> table;
    int wlen;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(N, blacklist);
 * int param_1 = obj.pick();
 */

File: 0701-0800/0710-Random Pick with Blacklist/0710-Random Pick with Blacklist.py
import random
class Solution:

    def __init__(self, N: int, blacklist: List[int]):
        self.wlen = N - len(blacklist)
        self.table = {b: -1 for b in blacklist}
        mapped = N - 1
        for b in blacklist:
            if b < self.wlen:
                while mapped in self.table:
                    mapped -= 1
                self.table[b] = mapped
                mapped -= 1
                

    def pick(self) -> int:
        index = random.randrange(self.wlen)
        return self.table.get(index, index)


# Your Solution object will be instantiated and called as such:
# obj = Solution(N, blacklist)
# param_1 = obj.pick()
